### PR TITLE
Perform final optimization at the end

### PR DESCRIPTION
**Pull request by ojura**
@damonkohler here's a small follow-up to #480 based on what you said there. There is now a clear distinction between finishing the trajectory and running the final optimization in `Node`. The offline node marks the trajectories finished as soon as it reaches the end of the respective bag, while the final optimization is run at the very end.

### CODE REVIEW

File: cartographer_ros/cartographer_ros/node.cc [hunk 1]
```diff
@@ -428,26 +427,35 @@ bool Node::HandleWriteState(
...
   }
-  map_builder_bridge_.RunFinalOptimization();
 }
 
 void Node::FinishTrajectory(const int trajectory_id) {
+  carto::common::MutexLocker lock(&mutex_);
+  FinishTrajectoryUnderLock(trajectory_id);
+}
+
+void Node::FinishTrajectoryUnderLock(const int trajectory_id) {
```
Comment (wohe) [inline cartographer_ros/cartographer_ros/node.cc]: Remove this convenience function? It does not seem to improve readability.

File: cartographer_ros/cartographer_ros/offline_node_main.cc [hunk 2]
```diff
@@ -209,16 +209,15 @@ void Run(const std::vector<string>& bag_filenames) {
...
-    // which might take a while.
-    clock_republish_timer.start();
     node.FinishTrajectory(trajectory_id);
-    clock_republish_timer.stop();
   }
 
-  // Republish the clock after bag processing has been completed.
+  // Ensure the clock is republished during the final optimization, which might
+  // take a while.
   clock_republish_timer.start();
```
Comment (wohe) [inline cartographer_ros/cartographer_ros/offline_node_main.cc]: Remove empty line?

File: cartographer_ros/cartographer_ros/offline_node_main.cc [hunk 3]
```diff
@@ -209,16 +209,15 @@ void Run(const std::vector<string>& bag_filenames) {
...
     bag.close();
-    // Ensure the clock is republished also during trajectory finalization,
-    // which might take a while.
-    clock_republish_timer.start();
     node.FinishTrajectory(trajectory_id);
-    clock_republish_timer.stop();
   }
 
-  // Republish the clock after bag processing has been completed.
+  // Ensure the clock is republished during the final optimization, which might
```
Comment (wohe) [inline cartographer_ros/cartographer_ros/offline_node_main.cc]: This is not the only purpose, right? What about serialization and the `--keep_running` flag?

File: cartographer_ros/cartographer_ros/node.cc (hunk 1 above)
Comment (ojura) [inline cartographer_ros/cartographer_ros/node.cc]: Sure, I can revert the last commit. I went by the rule "three or more, use a helper function". :-)

Comment (ojura) [inline cartographer_ros/cartographer_ros/node.cc]: I have reverted the commit which introduced the helper function.

**Pull request merged**